WWDC/DownloadManager.swift

```diff
@@ -12,69 +12,61 @@ import ConfCore
-    static let DownloadManagerDownloadFailed = Notification.Name("DownloadManagerDownloadFailed")
-    static let DownloadManagerDownloadFinished = Notification.Name("DownloadManagerDownloadFinished")
-    static let DownloadManagerDownloadProgressChanged = Notification.Name("DownloadManagerDownloadProgressChanged")
-
-}
-
 enum DownloadStatus {
     case none
-    case downloading(Double)
-    case paused
+    case downloading(DownloadManager.DownloadInfo)
+    case paused(DownloadManager.DownloadInfo)
     case cancelled
     case finished
     case failed(Error?)
 }
 
-extension URL {
-
-    var isDirectory: Bool {
-        guard isFileURL else { return false }
-        var directory: ObjCBool = false
-        return FileManager.default.fileExists(atPath: path, isDirectory: &directory) ? directory.boolValue : false
-    }
-
-    var subDirectories: [URL] {
-        guard isDirectory else { return [] }
-        return (try? FileManager.default.contentsOfDirectory(at: self, includingPropertiesForKeys: nil, options: [.skipsHiddenFiles]).filter { $0.isDirectory }) ?? []
-    }
-
-}
-
 final class DownloadManager: NSObject {
 
+    // Changing this dynamically isn't supported. Delete all downloads when switching
+    // from one quality to another otherwise you'll encounter minor unexpected behavior
+    static let downloadQuality = SessionAssetType.hdVideo
+
     private let log = OSLog(subsystem: "WWDC", category: "DownloadManager")
     private let configuration = URLSessionConfiguration.background(withIdentifier: "WWDC Video Downloader")
     private var backgroundSession: Foundation.URLSession!
-    private var downloadTasks: [String: URLSessionDownloadTask] = [:]
+    private var downloadTasks: [String: Download] = [:] {
+        didSet {
+            downloadTasksSubject.onNext(Array(downloadTasks.values))
+        }
+    }
+    private let downloadTasksSubject = BehaviorSubject<[Download]>(value: [])
+    var downloadsObservable: Observable<[Download]> {
+        return downloadTasksSubject.asObservable()
+    }
     private let defaults = UserDefaults.standard
 
-    private var storage: Storage!
+    var storage: Storage!
 
     static let shared: DownloadManager = DownloadManager()
 
     override init() {
         super.init()
 
-        backgroundSession = URLSession(configuration: configuration, delegate: self, delegateQueue: OperationQueue.main)
+        backgroundSession = URLSession(configuration: configuration, delegate: self, delegateQueue: .main)
     }
 
+    // MARK: - Session-based Public API
+
     func start(with storage: Storage) {
         self.storage = storage
 
         backgroundSession.getTasksWithCompletionHandler { _, _, pendingTasks in
             for task in pendingTasks {
-                if let key = task.originalRequest?.url!.absoluteString {
-                    self.downloadTasks[key] = task
+                if let key = task.originalRequest?.url!.absoluteString,
```

Comment:
I know `url` is probably never going to be nil, but just in case... ¯\_(ツ)_/¯ 

```suggestion
                if let key = task.originalRequest?.url?.absoluteString,
```